packages/core/core/src/ContentGraph.js

```diff
@@ -0,0 +1,84 @@
+// @flow strict-local
+
+import Graph, {type GraphOpts} from './Graph';
+import type {ContentKey, Node, NodeId} from './types';
+import nullthrows from 'nullthrows';
+
+export type SerializedContentGraph<
+  TNode: Node,
+  TEdgeType: string | null = null,
+> = {|
+  ...GraphOpts<TNode, TEdgeType>,
+  _contentKeyToNodeId: Map<ContentKey, NodeId>,
+|};
+
+export default class ContentGraph<
+  TNode: Node,
+  TEdgeType: string | null = null,
+> extends Graph<TNode, TEdgeType> {
+  _contentKeyToNodeId: Map<ContentKey, NodeId>;
+
+  constructor(opts: ?SerializedContentGraph<TNode, TEdgeType>) {
+    if (opts) {
+      let {_contentKeyToNodeId, ...rest} = opts;
+      super(rest);
+      this._contentKeyToNodeId = _contentKeyToNodeId;
+    } else {
+      super();
+      this._contentKeyToNodeId = new Map();
+    }
+  }
+
+  // $FlowFixMe[prop-missing]
+  static deserialize(
+    opts: SerializedContentGraph<TNode, TEdgeType>,
+  ): ContentGraph<TNode, TEdgeType> {
+    return new ContentGraph(opts);
+  }
+
+  // $FlowFixMe[prop-missing]
+  serialize(): SerializedContentGraph<TNode, TEdgeType> {
+    return {
+      ...super.serialize(),
+      _contentKeyToNodeId: this._contentKeyToNodeId,
+    };
+  }
+
+  addNodeByContentKey(contentKey: ContentKey, node: TNode): NodeId {
```

Comment:
Isn't `content === node.id` always true? So I think the first argument could be removed

Reply:
I believe we want to eventually remove id from Node since the id on node is the node value's id. @wbinnssmith , that's what we talked through, right?